File: civitasJAVA/civitas/src/civitas/SorpresaSalirCarcel.java
package civitas;

import java.util.ArrayList;


public class SorpresaSalirCarcel extends Sorpresa {

    MazoSorpresas mazo;

    //Constructor
    SorpresaSalirCarcel(MazoSorpresas mazo, String texto)
    {
        this.mazo = new MazoSorpresas();
        this.texto = "Esta sorpresa te lleva a la cárcel";

    }

    //Aplicar sorpresa a jugador pasado como parámetro
    void aplicarAJugador(int actual, ArrayList<Jugador> todos)
    {
        if (this.jugadorCorrecto(actual, todos)) {

            this.informe(actual, todos);
            todos.get(actual).obtenerSalvoconducto(new SorpresaSalvoconducto(mazo));
            salirDelMazo();
        }
    }

    //Devuelve si el jugador contiene irregularidades
    public boolean jugadorCorrecto(int actual, ArrayList<Jugador> todos)
    {
        return (actual < todos.size());
    }

    //Deshabilita propiedad especial
    void salirDelMazo()
    {
        mazo.inhabilitarCartaEspecial(this);
    }

    //Habilita propiedad especial
    void usada()
    {
        mazo.habilitarCartaEspecial(this);
    }

    //toString
    @Override
    public String toString() {
        return ("Salir Cárcel");
    }

}

File: civitasRUBY/lib/sorpresa_salircarcel.rb
#encoding: utf-8

require_relative "sorpresa"

module Civitas
  class SorpresaSalircarcel < Sorpresa
    def initialize(mazo)
      super("SaliCarcel")
      @mazo = mazo
    end
  
    #AplicarAJugador de SorpresaSalircarcel
    def aplicarAJugador(actual, todos)
      if (jugadorCorrecto(actual, todos))
        informe(actual, todos)

        haySalvoconducto = false        
        for jugador in todos
          if (jugador.tieneSalvoconducto())
            haySalvoconducto = true
          end
        end

        if (!haySalvoconducto)
          todos[actual].obtenerSalvoconducto(self)

        end
      end
    end
  
    #Inhabilita la carta especial en el mazo
    def salirDelMazo
      @mazo.inhabilitarCartaEspecial(self)
    end
  
    #Habilita la carta especial en el mazo
    def usada
      @mazo.habilitarCartaEspecial(self)
    end
    
    #toString de SorpresaSalircarcel
    def toString
      return "Sorpresa tipo SalirCarcel("+self.to_s+")"
    end
    
  end
end